File: flask/src/controller/user_controller.py
from flask import request
from response.response import response
from response.response_error import ResponseError
from service.user_service import get_users, get_user_by_id, create_user, get_user_saved_news, get_user_saved_news_comments, get_user_facilities, update_user

# Controller untuk mengambil semua data pengguna
def get_users_controller():
    try:
        page = request.args.get('page', default=1, type=int)
        limit = request.args.get('limit', default=100, type=int)

        result = get_users(page, limit)
        return response(200, result, "Get users success")
    except ResponseError as e:
        raise e
        
# Controller untuk mengambil data pengguna berdasarkan ID
def get_user_by_id_controller(id):
    try:
        result = get_user_by_id(id)
        return response(200, result, "Get user by id success")
    except ResponseError as e:
        raise e

# Controller untuk membuat pengguna baru
def create_user_controller():
    try:
        data = request.get_json()
        result = create_user(data)
        return response(200, result, "Create user success")
    except ResponseError as e:
        raise e
    
# Controller untuk mengambil berita yang disimpan oleh pengguna berdasarkan ID
def get_user_saved_news_controller(id):
    try:
        result = get_user_saved_news(id)
        return response(200, result, "Get user saved news success")
    except ResponseError as e:
        raise e
    
# Controller untuk mengambil komentar berita yang disimpan oleh pengguna berdasarkan ID
def get_user_saved_news_comments_controller(id):
    try:
        result = get_user_saved_news_comments(id)
        return response(200, result, "Get user saved news comments success")
    except ResponseError as e:
        raise e
    
# Controller untuk mengambil fasilitas yang dimiliki pengguna berdasarkan ID
def get_user_facilities_controller(id):
    try:
        result = get_user_facilities(id)
        return response(200, result, "Get user facilities success")
    except ResponseError as e:
        raise e
    
# Controller untuk memperbarui data pengguna berdasarkan ID
def update_user_controller(id):
    try:
        data = request.get_json()
        result = update_user(id, data)
        return response(200, result, "Update user success")
    except ResponseError as e:
        raise e

File: express/src/controller/user-controller.js
import userService from "../service/user-service.js";
import { response } from "../response/response.js";
import jwt from "jsonwebtoken";

// Controller untuk test, hanya mengembalikan data statis
const test = async (req, res, next) => {
  try {
    res.status(200).json({
      data: "test",
    });
  } catch (error) {
    next(error);
  }
};

// Controller untuk memverifikasi token JWT dan mengembalikan decoded data
const testToken = (req, res, next) => {
  try {
    const token = req.get("Authorization");
    const secret = process.env.JWT_SECRET;
    const decoded = jwt.verify(token, secret);
    res.status(200).json({
      data: decoded,
    });
  } catch (error) {
    next(error);
  }
};

// Controller untuk mendapatkan semua pengguna
const getUsers = async (req, res, next) => {
  try {
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 100;

    const result = await userService.getUsers(page, limit);
    response(200, result, "Get users success", res);
  } catch (error) {
    next(error);
  }
};

// Controller untuk membuat pengguna baru
const createUser = async (req, res, next) => {
  try {
    const result = await userService.createUser(req.body);
    response(201, result, "Create user success", res);
  } catch (error) {
    next(error);
  }
};

// Controller untuk mendapatkan pengguna berdasarkan ID
const getUserById = async (req, res, next) => {
  try {
    const result = await userService.getUserById(req.params.id);
    response(200, result, "Get user success", res);
  } catch (error) {
    next(error);
  }
};

// Controller untuk mendapatkan berita yang disimpan oleh pengguna
const getSavedNews = async (req, res, next) => {
  try {
    const result = await userService.getSavedNews(req.params.id);
    response(200, result, "Get saved news success", res);
  } catch (error) {
    next(error);
  }
};

// Controller untuk mendapatkan fasilitas yang dimiliki pengguna
const getFacilities = async (req, res, next) => {
  try {
    const result = await userService.getFacilities(req.params.id);
    response(200, result, "Get facilities success", res);
  } catch (error) {
    next(error);
  }
};

// Controller untuk memperbarui data pengguna berdasarkan ID
const updateUser = async (req, res, next) => {
  try {
    const result = await userService.updateUser(req.params.id, req.body);
    response(200, result, "Update user success", res);
  } catch (error) {
    next(error);
  }
};

// Controller untuk mendapatkan komentar berita yang disimpan oleh pengguna
const getSavedNewsComment = async (req, res, next) => {
  try {
    const result = await userService.getSavedNewsComment(req.params.id);
    response(200, result, "Get saved news comment success", res);
  } catch (error) {
    next(error);
  }
};

export default {
  test,
  getUsers,
  createUser,
  getUserById,
  getSavedNews,
  getFacilities,
  updateUser,
  getSavedNewsComment,
  testToken,
};
